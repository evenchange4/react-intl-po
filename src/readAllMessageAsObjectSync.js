import fs from 'fs';
import { sync as globSync } from 'glob';
import mergeWith from 'lodash/mergeWith';
import isArray from 'lodash/isArray';

const concatCustomizer = (accValue, objectValue) => {
  if (!isArray(accValue)) return objectValue;

  return accValue.concat(objectValue);
};

const mergeCustomizer = (accValue, objectValue) =>
  mergeWith(accValue, objectValue, concatCustomizer);

// hint: Use defaultMessage as key by default
const indexBy = (messageKey, messageContext) => ({ messages, filename }) =>
  messages.reduce(
    (acc, message) => ({
      ...acc,
      [message[messageKey]]: mergeWith(
        acc[message[messageKey]],
        {
          [messageContext ? message[messageContext] : '']: [
            { ...message, filename },
          ],
        },
        concatCustomizer,
      ),
    }),
    {},
  );

/**
 * Read extracted .json file synchronized and
 * aggregates origin messages objects
 *
 * @param {String} srcPatterns - path to translated .json file
 * @param {String} messageKey - [defaultMessage]
 * @return {Object} messages - return aggregates object
 *
 * @author Michael Hsu
 */

function readAllMessageAsObjectSync(
  srcPatterns,
  messageKey = 'defaultMessage',
  messageContext = '',
) {
  return (
    globSync(srcPatterns)
      // 1. read messages
      .map(filename => ({
        filename,
        messages: JSON.parse(fs.readFileSync(filename, 'utf8')),
      }))
      // 2. convert message list to nested objects by messageKey and messageContext
      .map(indexBy(messageKey, messageContext))
      // 3. aggregate objects (merge and concat)
      .reduce((acc, object) => mergeWith(acc, object, mergeCustomizer), {})
  );
}

export default readAllMessageAsObjectSync;
