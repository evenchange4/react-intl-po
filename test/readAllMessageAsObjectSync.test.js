import readAllMessageAsObjectSync from '../src/readAllMessageAsObjectSync';

it('should return a function', () => {
  expect(typeof readAllMessageAsObjectSync).toBe('function');
});

it('should return messages object with default messageKey', () => {
  expect(
    readAllMessageAsObjectSync('./test/messages/**/*.json'),
  ).toMatchSnapshot();
});

it('should return messages object with description as key', () => {
  expect(
    readAllMessageAsObjectSync('./test/messages/**/App.json', 'description'),
  ).toMatchSnapshot();
});

it('should return messages object with id as key', () => {
  expect(
    readAllMessageAsObjectSync('./test/messages/**/*.json', 'id'),
  ).toMatchSnapshot();
});

it('should return messages object with id as context', () => {
  expect(
    readAllMessageAsObjectSync('./test/messages/**/*.json', undefined, 'id'),
  ).toMatchSnapshot();
});
