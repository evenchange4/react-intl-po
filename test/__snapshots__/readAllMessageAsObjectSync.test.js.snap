// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return messages object with default messageKey 1`] = `
{
  "The device is now private or deleted.": {
    "": [
      {
        "id": "App.errorMessage",
        "description": "The error message when api response as 404 not found",
        "defaultMessage": "The device is now private or deleted.",
        "filename": "./test/messages/src/containers/App/App.json"
      }
    ]
  },
  "Go to MCS website": {
    "": [
      {
        "id": "App.errorButton",
        "description": "Click error Button",
        "defaultMessage": "Go to MCS website",
        "filename": "./test/messages/src/containers/App/App.json"
      },
      {
        "id": "NotFound.errorButton",
        "description": "Click error Button",
        "defaultMessage": "Go to MCS website",
        "filename": "./test/messages/src/containers/NotFound/messages.json"
      }
    ]
  },
  "Creator": {
    "": [
      {
        "id": "App.Creator 1",
        "description": "Creator 1",
        "defaultMessage": "Creator",
        "filename": "./test/messages/src/containers/App/App.json"
      },
      {
        "id": "App.Creator 2",
        "description": "Creator 2",
        "defaultMessage": "Creator",
        "filename": "./test/messages/src/containers/App/App.json"
      },
      {
        "id": "NotFound.Creator",
        "description": "Creator",
        "defaultMessage": "Creator",
        "filename": "./test/messages/src/containers/NotFound/messages.json"
      }
    ]
  }
}
`;

exports[`should return messages object with description as key 1`] = `
{
  "The error message when api response as 404 not found": {
    "": [
      {
        "id": "App.errorMessage",
        "description": "The error message when api response as 404 not found",
        "defaultMessage": "The device is now private or deleted.",
        "filename": "./test/messages/src/containers/App/App.json"
      }
    ]
  },
  "Click error Button": {
    "": [
      {
        "id": "App.errorButton",
        "description": "Click error Button",
        "defaultMessage": "Go to MCS website",
        "filename": "./test/messages/src/containers/App/App.json"
      }
    ]
  },
  "Creator 1": {
    "": [
      {
        "id": "App.Creator 1",
        "description": "Creator 1",
        "defaultMessage": "Creator",
        "filename": "./test/messages/src/containers/App/App.json"
      }
    ]
  },
  "Creator 2": {
    "": [
      {
        "id": "App.Creator 2",
        "description": "Creator 2",
        "defaultMessage": "Creator",
        "filename": "./test/messages/src/containers/App/App.json"
      }
    ]
  }
}
`;

exports[`should return messages object with id as context 1`] = `
{
  "The device is now private or deleted.": {
    "App.errorMessage": [
      {
        "id": "App.errorMessage",
        "description": "The error message when api response as 404 not found",
        "defaultMessage": "The device is now private or deleted.",
        "filename": "./test/messages/src/containers/App/App.json"
      }
    ]
  },
  "Go to MCS website": {
    "App.errorButton": [
      {
        "id": "App.errorButton",
        "description": "Click error Button",
        "defaultMessage": "Go to MCS website",
        "filename": "./test/messages/src/containers/App/App.json"
      }
    ],
    "NotFound.errorButton": [
      {
        "id": "NotFound.errorButton",
        "description": "Click error Button",
        "defaultMessage": "Go to MCS website",
        "filename": "./test/messages/src/containers/NotFound/messages.json"
      }
    ]
  },
  "Creator": {
    "App.Creator 1": [
      {
        "id": "App.Creator 1",
        "description": "Creator 1",
        "defaultMessage": "Creator",
        "filename": "./test/messages/src/containers/App/App.json"
      }
    ],
    "App.Creator 2": [
      {
        "id": "App.Creator 2",
        "description": "Creator 2",
        "defaultMessage": "Creator",
        "filename": "./test/messages/src/containers/App/App.json"
      }
    ],
    "NotFound.Creator": [
      {
        "id": "NotFound.Creator",
        "description": "Creator",
        "defaultMessage": "Creator",
        "filename": "./test/messages/src/containers/NotFound/messages.json"
      }
    ]
  }
}
`;

exports[`should return messages object with id as key 1`] = `
{
  "App.errorMessage": {
    "": [
      {
        "id": "App.errorMessage",
        "description": "The error message when api response as 404 not found",
        "defaultMessage": "The device is now private or deleted.",
        "filename": "./test/messages/src/containers/App/App.json"
      }
    ]
  },
  "App.errorButton": {
    "": [
      {
        "id": "App.errorButton",
        "description": "Click error Button",
        "defaultMessage": "Go to MCS website",
        "filename": "./test/messages/src/containers/App/App.json"
      }
    ]
  },
  "App.Creator 1": {
    "": [
      {
        "id": "App.Creator 1",
        "description": "Creator 1",
        "defaultMessage": "Creator",
        "filename": "./test/messages/src/containers/App/App.json"
      }
    ]
  },
  "App.Creator 2": {
    "": [
      {
        "id": "App.Creator 2",
        "description": "Creator 2",
        "defaultMessage": "Creator",
        "filename": "./test/messages/src/containers/App/App.json"
      }
    ]
  },
  "NotFound.errorButton": {
    "": [
      {
        "id": "NotFound.errorButton",
        "description": "Click error Button",
        "defaultMessage": "Go to MCS website",
        "filename": "./test/messages/src/containers/NotFound/messages.json"
      }
    ]
  },
  "NotFound.Creator": {
    "": [
      {
        "id": "NotFound.Creator",
        "description": "Creator",
        "defaultMessage": "Creator",
        "filename": "./test/messages/src/containers/NotFound/messages.json"
      }
    ]
  }
}
`;
